ipamd/ipamd.go

```diff
@@ -348,177 +352,254 @@ func (c *IPAMContext) StartNodeIPPoolManager() {
 }
 
 func (c *IPAMContext) updateIPPoolIfRequired() {
-	c.retryAllocENIIP()
 	if c.nodeIPPoolTooLow() {
 		c.increaseIPPool()
 	} else if c.nodeIPPoolTooHigh() {
-		c.decreaseIPPool()
+		c.decreaseIPPool(decreaseIPPoolInterval)
 	}
 }
 
-// TODO: Does not retry!
-func (c *IPAMContext) retryAllocENIIP() {
-	ipamdActionsInprogress.WithLabelValues("retryAllocENIIP").Add(float64(1))
-	defer ipamdActionsInprogress.WithLabelValues("retryAllocENIIP").Sub(float64(1))
+func (c *IPAMContext) decreaseIPPool(interval time.Duration) {
+	ipamdActionsInprogress.WithLabelValues("decreaseIPPool").Add(float64(1))
+	defer ipamdActionsInprogress.WithLabelValues("decreaseIPPool").Sub(float64(1))
 
-	curIPTarget, warmIPTargetDefined := c.getCurWarmIPTarget()
-	if warmIPTargetDefined && curIPTarget <= 0 {
-		log.Debugf("Skipping retry allocating ENI IP, warm IP target reached")
+	now := time.Now()
+	timeSinceLast := now.Sub(c.lastDecreaseIPPool)
+	if timeSinceLast <= interval {
+		log.Debugf("decreaseIPPool: skipping because time since last %v <= %v", timeSinceLast, interval)
 		return
 	}
-	maxIPLimit, err := c.awsClient.GetENIipLimit()
-	if err != nil {
-		log.Infof("Failed to retrieve ENI IP limit: %v", err)
-		return
-	}
-	eni := c.dataStore.GetENINeedsIP(maxIPLimit, UseCustomNetworkCfg())
-	if eni != nil {
-		log.Debugf("Attempt again to allocate IP address for ENI :%s", eni.ID)
-		var err error
-		if warmIPTargetDefined {
```

Comment:
See line 587 to see where the IP allocation logic moved for cases where ENIs should have more assigned IPs than they currently have.